src/exec/mod.rs

```diff
@@ -196,14 +198,20 @@ impl CommandBuilder {
     fn finish(&mut self) -> io::Result<()> {
         if self.count > 0 {
             self.cmd.try_args(&self.post_args)?;
-            self.cmd.status()?;
+            if !self.cmd.status()?.success() {
+                self.exit_code = ExitCode::GeneralError;
```

Comment:
Rather than storing an exit code here, I think we should return an `Err`.

Reply:
Hmm, my idea went like this: We *do not* want to short-circuit on non-success exit codes (compare `-x` semantics), so we were to return an `Err`, we would have to catch it (but just this one, not other `Err`s) in `execute_batch` and still maintain sth like this `exit_code` there. But this seemed just more convoluted, so I took this approach.

Happy to change though if you have a different idea :+1:

Reply:
The way `-x` works is if a command fails, we short circuit for that particular path.

So for example if you have `fd foo -x c1  \; -x c2`   and c1 fails for a specific path, then c2 won't be run for that path. But it will continue trying to run c1 for other paths 

I think we would probably want similar behavior for exec-batch. If one command fails, we probably don't want to run the next one in the series, because it might depend on the first one succeeding.

However, maybe we do want to run the commandset with the next batch of paths if we have more paths than can fit in a single command invocation?

I created #1138  which will short circuit.

Reply:
I was actually only thinking about batching in my comment above, thanks for mentioning multiple commands.

> However, maybe we do want to run the commandset with the next batch of paths if we have more paths than can fit in a single command invocation?

Yes, I think that we should run either all or no batches of any given command, to be consistent with `-x` (where one can be sure that a command is either executed for all files or for none).

---

So in short, it seems reasonable to have short circuiting for mulitple commands, but no short circuiting for multiple batches of the same command.

I can update this PR if you agree, but if short circuiting batches is prefered, then #1138 seems nicer :+1:

Reply:
I don't have a very strong preference on  whether or not we short circuit for batches for a single command. @sharkdp do you have any thoughts on it?

Reply:
`find` doesn't short circuit in this case, and I think that's the right behaviour.  Short circuiting is confusing since the commands don't necessarily get the same arguments.

Reply:
Okay, to summarize (mostly for me to make sure I understand it correctly):

Let's say we execute
```
fd --batch-size=2 -X cmdA \; -X cmdB
```
and there are five search results `1`, `2`, …, `5`. Then if everything goes well (both commands always succeed), we run the following commands (modulo ordering):
```
cmdA ./1 ./2
cmdB ./1 ./2
cmdA ./3 ./4
cmdB ./3 ./4
cmdA ./5
cmdB ./5
```

Now the question is what happens if `cmdA` fails for the first batch (`cmdA ./1 ./2`). There are three reasonable options:
1. Quit immediately.
2. Only continue to run `cmdA` on the remaining batches (`cmdA ./3 ./4` and `cmdA ./5`). Do not run `cmdB` on any batch.
3. Still run everything as if nothing happened.

> I think we would probably want similar behavior for exec-batch. If one command fails, we probably don't want to run the next one in the series, because it might depend on the first one succeeding.
> 
> However, maybe we do want to run the commandset with the next batch of paths if we have more paths than can fit in a single command invocation?

@tmccombs I understand you would like to advocate for option 2?

> `find` doesn't short circuit in this case, and I think that's the right behaviour. Short circuiting is confusing since the commands don't necessarily get the same arguments.

@tavianator It wasn't clear to me what you meant by "in this case". But if my experiments are correct, `find` does not short-circuit at all, neither with respect to batches, nor with respects to follow-up commands. This means `find` implements option 3.


If my interpretation is correct, I'm also voting for `find`-compatibility with option 3. However, I'm not sure I understand your reasoning here:

> Short circuiting is confusing since the commands don't necessarily get the same arguments.

Reply:
> @tavianator It wasn't clear to me what you meant by "in this case"

I just meant that `find -exec ... {} +` does no short circuiting, as opposed to `find -exec ... {} \;`

> If my interpretation is correct, I'm also voting for find-compatibility with option 3. However, I'm not sure I understand your reasoning here:
>
> > Short circuiting is confusing since the commands don't necessarily get the same arguments.

What I mean is that in general, you cannot assume `cmdA` has run for a particular file before `cmdB`, as the batches may end up different due to command line length limits.  E.g. you may get

```
cmdB ./1 ./2 ./3
cmdA ./1 ./2 ./3 ./4
cmdA ./5
cmdB ./4 ./5
```

or something, especially if `cmdB` is longer than `cmdA`.

Since you can't rely on it in general, I don't think we should offer it on a best-effort basis either, so that people don't rely on it accidentally.

Reply:
Ah.... I didn't think about that. Now I think 3 might be the right way to go, at least for batches. 

Or possibly we change the way batches work, and we run all batches for each command before moving on to the next one, but that could hurt performance.

Reply:
> Or possibly we change the way batches work, and we run all batches for each command before moving on to the next one, but that could hurt performance.

That's not a bad idea, but it's probably material for another PR :).  I don't think the perf impact would be too bad

Reply:
Okay. So how do we proceed with this PR? I think it would be great to add some tests for the batch-behavior as well. We can use `--batch-size=N` for that.

Reply:
So I understand it correctly that the semantics of this PR (no short circuiting at all, i.e. @sharkdp option 3.) are good for now? In that case, I will add a few tests ensuring that all commands are executed even if only some commands/batches fail :+1:

Reply:
@amesgen That would be great